doc/akfavatar-term.de.man
.\" Process this file with
.\" groff -man -Tutf8 akfavatar-term.de.man
.\"
.
.\" Macros .TQ .EX .EE taken from groff an-ext.tmac
.\" Copyright (C) 2007, 2009 Free Software Foundation, Inc.
.\" You may freely use, modify and/or distribute this file.
.
.\" Continuation line for .TP header.
.de TQ
.  br
.  ns
.  TP \\$1\" no doublequotes around argument!
..
.
.\" Start example.
.de EX
.  nr mE \\n(.f
.  nf
.  nh
.  ft CW
..
.
.
.\" End example.
.de EE
.  ft \\n(mE
.  fi
.  hy \\n(HY
..
.
.TH "akfavatar-term" 3 2012-07-21 AKFAvatar
.nh
.
.SH NAME
akfavatar-term \- Modul zum Ausf\[:u]hren von Terminal-basierten Programmen in AKFAvatar
.
.SH SYNTAX
.PP
term = require "akfavatar-term"
.PP
.BR Hinweis :
Man muss eine globale Variable verwenden, wenn man die APC-Schnittstelle
verwenden will.
.PP
.SH BESCHREIBUNG
.PP
Dieses Modul kann man verwenden um allgemeine Terminal-basierte Programme
innerhalb von AKFAvatar laufen zu lassen.
Die Terminal-Emulation emuliert eine Linux-Textkonsole
(nicht nur auf Linux-basierten Systemen).
.PP
Es kann auch daf\[:u]r verwendet werden, um spezifische Programme f\[:u]r AKFAvatar zu
schreiben, in Programmiersprachen, die ansonsten nicht unterst\[:u]tzt werden.
Es gibt ein paar M\[:o]glichkeiten mit denen das ausf\[:u]hrende Skript und das
auszuf\[:u]hrende Programm miteinander kommunizieren k\[:o]nnen.
.PP
.BR Achtung :
Dieses Modul steht nicht f\[:u]r allen Plattformen zur Verf\[:u]gung!
Insbesondere nicht f\[:u]r Windows.
.PP
.TP
.BI term.startdir( [Verzeichnis] )
Legt das Anfangsverzeichnis f\[:u]r den n\[:a]chsten
.BR term.execute() -Befehl
fest.
Wenn kein
.I Verzeichnis
angegeben wird, wird die vorherige Einstellung dadurch gel\[:o]scht.
.IP
.BR Hinweis :
Das Arbeitsverzeichnis des ausf\[:u]hrenden Skriptes und des ausgef\[:u]hrten
Programmes sind voneinander unabh\[:a]ngig!
.PP
.TP
.B term.homedir()
Setzt das Anfangsverzeichnis f\[:u]r den n\[:a]chsten
.BR term.execute() -Befehl
auf das Heimatverzeichnis des Benutzers.
.IP
.BR Hinweis :
Das Arbeitsverzeichnis des ausf\[:u]hrenden Skriptes und des ausgef\[:u]hrten
Programmes sind voneinander unabh\[:a]ngig!
.PP
.TP
.BI term.color( "true|false" )
Schaltet die Farbunterst\[:u]tzung ein oder aus.
Der Terminal-Typ mit Farben ist \[Bq]linux\[lq],
ohne Farben \[Bq]linux-m\[lq].
.PP
.TP
.BI term.setenv( "Variable, Wert" )
Setzt die angegebene
.RI Umgebungs variable
auf den angegebenen
.IR Wert .
.br
Das hat auch Einfluss auf Programme, die mit
.BR os.execute() " oder " io.popen()
ausgef\[:u]hrt werden.
.PP
.TP
.BI term.unsetenv( Variable )
L\[:o]scht die angegebene
.RI Umgebungs variable .
.br
Das hat auch Einfluss auf Programme, die mit
.BR os.execute() " oder " io.popen()
ausgef\[:u]hrt werden.
.PP
.TP
.BI term.execute( "[Programm [, Argumente]]" )
F\[:u]hrt das angegebene
.I Programm
mit den angegebenen
.IR Argumente n
in der Terminal-Emulation aus.
Wenn kein
.I Programm
angegeben wird, wird die Standard-Shell ausgef\[:u]hrt.
.PP
.SS APC-Schnittstelle
.PP
Die \[Bq]Application Program Command\[lq] (APC) Schnittstelle ist
eine Methode mit der das ausgef\[:u]hrte Programm mit dem Lua-Interpreter
des ausf\[:u]hrenden Programmes kommunizieren kann.
Um dies zu verwenden sendet das ausgef\[:u]hrte Programm Escape-Sequenzen der Art,
wie ESC + \[Bq]_\[lq], gefolgt vom Befehl, abgeschlossen mit 
ESC + \[Bq]\\\[lq].
Die Ausgabe muss eventuell noch einen Flush bekommen, um einen sofortigen
Effekt zu erzielen.
.PP
Zum Beispiel in C:
.PP
.EX
static void
APC (const char *s)
{
  fprintf (stdout, "\\033_%s\\033\\\\", s);
  fflush (stdout);
}
.EE
.PP
Die Befehle k\[:o]nnen Lua-Funktionsaufrufe sein, insbesondere die 
\[Bq]avt.\[lq]-Befehle von \[Bq]lua-akfavatar\[lq]
(siehe \[Bq]lua-akfavatar-ref\[lq]).
.br
Bis zu 1024 Zeichen k\[:o]nnen gesendet werden.
Das bedeutet, man kann keinen komplizierten Code vom ausgef\[:u]hrten Programm
senden.
Allerdings kann man im Lua-Skript, das das Programm aufruft globale Funktionen
definieren, die dann auch mit der APC-Schnittstelle aufrufbar sind.
.PP
Als Konvention sollte das aufrufende Skript eine Umgebungsvariable namens
.I "\[Bq]APC\[lq]"
definieren, die die verf\[:u]gbaren Funktionen beschreibt.
.PP
Zum Beispiel:
.PP
.EX
  term.setenv("APC", _VERSION .. ", lua-akfavatar")
.EE
.PP
.TP
.BI term.send( String )
Sendet einen String an das ausgef\[:u]hrte Programm, so als ob es von der
Tastatur eingegeben worden w\[:a]re.
Das kann man nur von der APC-Schnittstelle (siehe oben) verwenden.
.br
Wenn das ausgef\[:u]hrte Programm zeilenorientiert ist, sollte der String mit
einem \[Bq]\\r\[lq] (wie Return) abgeschlossen werden.
.PP
.TP
.BI term.decide( "String1 [,String2]" )
Zeige Plus-/Minus-Kn\[:o]pfe und abh\[:a]ngig von dieser Auswahl sende
.IR String1 " oder " String2
an das ausgef\[:u]hrte Programm, so als ob es von Tastatur eingegeben
worden w\[:a]re.
Das kann man nur von der APC-Schnittstelle (siehe oben) verwenden.
.br
Wenn das ausgef\[:u]hrte Programm zeilenorientiert ist, sollten beide Strings
angegeben werden und beide mit einem \[Bq]\\r\[lq] (wie Return)
abgeschlossen werden.
.PP
.SH "SIEHE AUCH"
.BR lua-akfavatar (1)
.BR lua (1)
.BR lua-akfavatar-ref (3)
.BR akfavatar-graphic (3)
.BR akfavatar.utf8 (3)


NEWS
* AKFAvatar-testing

 - optimized printing of text (line buffered)
 - audio-data with same specs as former one is played with less latency
 - improved support for 24 and 32 bit audio
 - new file Bücher.lua for reading German books from my server (needs curl)

  C-API changes:
    - new macro: AVT_KEY_F

* AKFAvatar 0.24.3

  - audio: big audio files are read directly from file while playing
    (if possible via mmap)
  - audio: AU and WAV data can be read from non-seekable streams
    (loaded into memory)
  - fix: Pascal sound support fixed
  - the control characters SO/SI switch underlined mode on/off;
    the control characters DC1/DC2 switch bold mode on/off
  - configure: new option --with-sdl2
  - SDL2: support for SDL2 with automatic resizing of the graphics
  - SDL2: support for pasting text from clipboard (Alt+V, middle mouse button)

  Lua-API changes:
    - new functions: avt.frequency
    - module "akfavatar-term": fix APC interface

* AKFAvatar 0.24.2

  - lua/clock.lua: resizing works (with some delay)
  - lua/four_in_a_row.lua: mouse support / resizing
  - lua/four_in_a_row.lua: play against computer

  Lua-API changes:
    - Module "akfavatar-graphic":
      - new functions: graphic.set_pointer_buttons_key,
        graphic.set_pointer_motion_key, graphic.get_pointer_position,
        graphic.set_resize_key

* AKFAvatar 0.24.1

  - fix: lua/interactive_lua.lua: fix handling of incomplete lines
  - Lua module "akfavatar-graphic" included in executable
  - text cursor modified
  - lua-akfavatar: GNU .info files can be shown as plain text (no hyperlinks)
  - lua/hexviewer.lua: show more characters on the right side
  - lua/multiplication.lua: clear board (no scrolling)
  - configure accepts the option --with-linuxfb
    which then builds for the Linux framebuffer without the SDL
    (no mouse and no audio support yet)

  C-API changes:
    - new functions: avt_detect_utf8

  Lua-API changes:
    - new functions: avt.toutf8, avt.detect_utf8, avt.utf8codepoints

* AKFAvatar 0.24.0

  - fix: terminal doesn't use so much CPU time anymore
  - fix: window resizing for the pager works again
  - fix: switching fullscreen mode while avatar moves in or out
  - fix: avt_credits sometimes got the end wrong
  - fix: audioplayer.lua accepts PLS playlists again
  - fix: Pascal binding works again, extended
  - paging through menus is animated
  - lua/audioplayer.lua: improved playlist support
  - lua/multiplication.lua: more authentic

  C-API changes:
    - avt_char is unsigned int now, avt_choice adapted
      should be compatible with char32_t in C11
    - functions with *_mb* replaced with functions *_char*
    - new functions: avt_darker, avt_brighter, avt_brightness
    - avtadons.h also uses the prefix avt_ for everything

  Lua-API changes:
    - The default encoding is now ASCII
      to encourage always using avt.encoding()
    - avt.set_directory or avt.chdir return true on success
      or nil and an error message on error

  Pascal-API changes:
    - variable CheckEsc removed
    - new procedures/functions: Menu, ReadKeyCode, reserve_single_keys


* AKFAvatar 0.23.1

  - fixed vulnerabilities
  - avoid recursive calls of the bell
  - split source code for better maintainablitity
  - name sign with 3d effect

  C-API changes:
    - avt_set_title: no automatic charset conversion
    - avt_put_char: support for UTF-16 surrogate characters
    - new functions: avt_wide_copyright, avt_wide_license, avt_wide_version,
      avt_menu, avt_palette_size
    - avtaddons:
      - avta_file_selection and avta_filter have a new parameter "data",
        which is passed along

  Lua-API changes:
    - Module "akfavatar-graphic"
      - new methods: gr:border3d

* AKFAvatar 0.23.0

  The code was made less dependend on SDL. There is a new experimantal
  backend for a Linux framebuffer device, but without mouse support or audio
  (of course you can call an external audio-player in your scripts).

  - bugfix: fix for XPM images with more than 256 colors
  - bugfix: lua-akfavatar also resets the charset
  - bugfix/change: lua-akfavatar option -l sets a global variable
  - avt_update() is much faster
  - support for combining characters (Unicode)
  - The pager supports numbers (number-pad when NumLock is on)
  - lua/interactive_lua.lua supports scrolling back in a history list
  - regression: the navigation bar cannot handle the pause key anymore
  - key codes have been changed not to conflict with the "corporate zone"

  C-API changes:
    - all deprecated functions removed
    - also removed: avt_register_keyhandler, avt_register_mousehandler
    - avt_update() is much faster
    - avt_get_key replaces avt_key
    - modified: avt_show_raw_image, avt_put_raw_image_file,
      avt_put_raw_image_stream, avt_put_raw_image_data, avt_put_raw_image_xpm
      don't have the bpp parameter anymore, only 4 Bytes per pixel accepted
    - new functions: avt_key_pressed, avt_push_key, avt_clear_keys,
      avt_set_audio_end_key, avt_combining, avt_bell_function
    - new experimental functions: avt_set_pointer_buttons_key,
      avt_set_pointer_motion_key, avt_get_pointer_position,
      avt_char avt_input, avt_input_mb
    - new macro: avt_elapsed

  Lua-API changes:
    - the function avt.ask was extended (yet undocumented)
    - new functions: avt.combining, avt.key_pressed, avt.clear_keys,
      avt.push_key, avt.set_audio_end_key, avt.alert, avt.getcwd
    - new table: avt.key with key codes


* AKFAvatar 0.22.1

  There are a lot of changes in the APIs. Also the ABI was changed.

  - Window is resizable again (content gets centered)
  - The image can be a separated banner as header or footer
  - Uses Lua-5.2 now, no longer compatible with 5.1
  - Lua modules don't define globals anymore (scripts have to be adapted!)
  - Data files are put in a separate directory and Lua has a search function
    for it (scripts have to be adapted!)
  - the function for translations is integrated in Lua-AKFAvatar

lua/akfavatar-term.c
/*
 * AKFAvatar Terminal emulation for Lua 5.2
 * ATTENTION: this is work in progress, ie. not finished yet
 * Copyright (c) 2010,2011,2012,2013
 * Andreas K. Foerster <akf@akfoerster.de>
 *
 * This file is part of AKFAvatar
 *
 * AKFAvatar is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * AKFAvatar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

#define _ISOC99_SOURCE
#define _XOPEN_SOURCE 600

#include "akfavatar.h"
#include "avtaddons.h"

#include <stdlib.h>
#include <string.h>
#include <wchar.h>
#include <locale.h>
#include <iso646.h>

#include <sys/types.h>
#include <pwd.h>

#ifndef NO_LANGINFO
#  include <langinfo.h>
#endif

#ifdef __cplusplus
extern "C"
{
#endif

#include <lua.h>
#include <lauxlib.h>
#include <lualib.h>

  extern int luaopen_term (lua_State * L);

#ifdef __cplusplus
}
#endif


static char *startdir = NULL;
static lua_State *term_L;	// lua_State of the running terminal

static int
quit (lua_State * L)
{
  if (avt_get_status () <= AVT_ERROR)
    {
      return luaL_error (L, "%s", avt_get_error ());
    }
  else				// stop requested
    {
      // no actual error, so no error message
      // this is handled by the calling program
      lua_pushnil (L);
      return lua_error (L);
    }
}

/*
 * ececute program in terminal
 * accepts a program and its parameters as separate arguments
 * if no arguments are given, it starts the default shell
 */
static int
lterm_execute (lua_State * L)
{
  int fd;
  int n;
  char *argv[256];

  if (not avt_initialized ())
    luaL_error (L, "execute: AKFAvatar not initialized");

  // set all locale settings
  setlocale (LC_ALL, "");

  n = lua_gettop (L);		// number of options

  // number of arguments is already limited in Lua, but I want to be save
  if (n > 255)
    luaL_error (L, "execute: too many arguments");

  if (n >= 1)			// start program
    {
      for (int i = 0; i < n; i++)
	argv[i] = (char *) luaL_checkstring (L, i + 1);
      argv[n] = NULL;

      fd = avt_term_start (startdir, argv);
    }
  else				// start shell
    fd = avt_term_start (startdir, NULL);


  if (fd != -1)
    {
      term_L = L;
      avt_term_run (fd);
      term_L = NULL;
    }

  if (startdir)
    {
      free (startdir);
      startdir = NULL;
    }

  if (avt_get_status () != AVT_NORMAL)
    quit (L);

  return 0;
}

/*
 * set the start directory for the next execute command
 * none or nil as attribute cancels the directory
 */
static int
lterm_startdir (lua_State * L)
{
  if (startdir)
    {
      free (startdir);
      startdir = NULL;
    }

  if (not lua_isnoneornil (L, 1))
    startdir = strdup (luaL_checkstring (L, 1));

  return 0;
}

static int
lterm_homedir (lua_State * L)
{
  (void) L;

  char *home;

  home = getenv ("HOME");

  // when the variable is not set, dig deeper
  if (not home or not * home)
    {
      struct passwd *user_data;
      user_data = getpwuid (getuid ());	// POSIX.1-2001
      if (user_data and user_data->pw_dir and * user_data->pw_dir)
	home = user_data->pw_dir;
    }

  if (home)
    {
      if (startdir)
	free (startdir);

      startdir = strdup (home);
    }

  return 0;
}

static int
lterm_setenv (lua_State * L)
{
  setenv (luaL_checkstring (L, 1), luaL_checkstring (L, 2), 1);

  return 0;
}

static int
lterm_unsetenv (lua_State * L)
{
  unsetenv (luaL_checkstring (L, 1));

  return 0;
}

static int
lterm_color (lua_State * L)
{
  luaL_checktype (L, 1, LUA_TBOOLEAN);
  avt_term_nocolor (not lua_toboolean (L, 1));

  return 0;
}

// send string to stdin of process (APC), add "\r" for return
static int
lterm_send (lua_State * L)
{
  size_t len;
  const char *buf;

  if (not term_L)
    return
      luaL_error (L, "send: only for Application Program Commands (APC)");

  buf = luaL_checklstring (L, 1, &len);
  avt_term_send (buf, len);

  return 0;
}

/*
 * lets the user decide
 * sends string 1 or 2 if given to stdin of process
 */
static int
lterm_decide (lua_State * L)
{
  size_t len;
  const char *buf;

  if (not term_L)
    return
      luaL_error (L, "decide: only for Application Program Commands (APC)");

  if (avt_decide ())
    buf = lua_tolstring (L, 1, &len);
  else
    buf = lua_tolstring (L, 2, &len);

  if (buf)
    avt_term_send (buf, len);

  return 0;
}

/*
 * Application Program Commands (APC)
 * in this case Lua commands
 * up to 1024 chars
 */
static int
APC_command (wchar_t * command)
{
  if (not term_L or not command or not * command)
    return -1;

  const struct avt_charenc *convert = avt_char_encoding (NULL);

  size_t rest = 4097;
  char *mbstring = malloc (rest);
  if (not mbstring)
    return -1;

  char *restrict p = mbstring;
  wchar_t *restrict c = command;
  size_t len = 0;

  while (*c and rest > 0)
    {
      size_t l = convert->encode (convert, p, rest, *c);
      p += l;
      len += l;
      rest -= l;
      ++c;
    }

  *p = '\0';

  int ret = luaL_loadbufferx (term_L, mbstring, len, mbstring, "t");
  free (mbstring);

  if (ret != LUA_OK)
    return lua_error (term_L);

  lua_call (term_L, 0, 0);
  avt_term_update_size ();	// in case the size changed

  return 0;
}

static const luaL_Reg termlib[] = {
  {"startdir", lterm_startdir},
  {"homedir", lterm_homedir},
  {"color", lterm_color},
  {"setenv", lterm_setenv},
  {"unsetenv", lterm_unsetenv},
  {"execute", lterm_execute},
  {"send", lterm_send},
  {"decide", lterm_decide},
  {NULL, NULL}
};

int
luaopen_term (lua_State * L)
{
  term_L = NULL;
  luaL_newlib (L, termlib);
  avt_term_register_apc (APC_command);
  return 1;
}
